### PR TITLE
add x86 schedule injective, enhance simplify(min/max/mod/var...), enh…

### PR DESCRIPTION
**Pull request by wenming2014**
1、binary x86 schedule injective，e.g. add with shape[100,32] can gain 600+% op performance benefits
2、enhance simplify(min/max/mod/var...)
3、enhance vectorize loop、 ramp calculation、 buffer align alloc、 bool type and llvm codegen、add call attrs
4、fix some bugs

### CODE REVIEW

File: cinn/common/ir_util.cc [hunk 1]
```diff
@@ -49,8 +50,17 @@ Expr RampRelatedAdd(ir::Broadcast *broadcast, Expr other) {
...
 }
-// ramp * ramp
+// ramp + ramp
 Expr RampRelatedAdd(ir::Ramp *ramp, ir::Ramp *other) {
+  CHECK(ramp);
+  CHECK(other);
+  if (ramp->lanes == other->lanes) {
+    Expr base_add   = common::AutoSimplify(ramp->base + other->base);
+    Expr stride_add = common::AutoSimplify(ramp->stride + other->stride);
+    LOG(INFO) << base_add;
```
Comment (Superjomn) [inline cinn/common/ir_util.cc]: Change LOG to VLOG in module code.

LOG is only allowed in tests.

File: cinn/hlir/framework/buffer.h [hunk 2]
```diff
@@ -50,6 +54,11 @@ struct Buffer final {
     return memory_mng_cache_->malloc(size);
   }
 
+  inline void* Aligned_Alloc(uint32_t alignment, uint32_t size) CINN_RESULT_SHOULD_USE {
```
Comment (Superjomn) [inline cinn/hlir/framework/buffer.h]: AlignedAlloc ? 

Is the `_` necessary?

File: cinn/hlir/pe/schedule.cc [hunk 3]
```diff
@@ -1,7 +1,112 @@
...
+      return;
+    }
+    auto [j_outer, j_inner] = stage->Split(fused, split_factor);
+    stage->Vectorize(j_inner, split_factor);
+  }
+
+  return;
+}
+
+void CudaScheduleMul(poly::StageMap stages,
```
Comment (Superjomn) [inline cinn/hlir/pe/schedule.cc:43]: @haozech Is the CUDA schedule duplcate here?

File: cinn/ir/ir.h [hunk 4]
```diff
@@ -474,10 +479,11 @@ struct LoadStoreAddrMnger {
  */
 struct Load : public ExprNode<Load>, public LoadStoreAddrMnger {
   std::vector<Expr> indices;
+  Expr index_;
```
Comment (Superjomn) [inline cinn/ir/ir.h]: The `index_` should be a private member.

But IR node's member should be all public, or some state will differ after copy or visit.

File: tests/benchmark/test_all_ops_default.cc [hunk 5]
```diff
@@ -118,8 +142,8 @@ std::vector<std::vector<int>> shapes_pool2d1 = {{2, 1024, 14, 14}};
 TEST_DEFAULT1(pool2d, pool2d1, type, attr_store_pool2d)
 
 // softmax
-std::vector<std::vector<int>> shapes_softmax = {{1024, 2048}};
-TEST_DEFAULT(softmax, softmax, type1)
+// std::vector<std::vector<int>> shapes_softmax = {{1024, 2048}};
```
Comment (Superjomn) [inline tests/benchmark/test_all_ops_default.cc]: clean useless code

File: cinn/optim/vectorize_loops.cc [hunk 6]
```diff
@@ -331,17 +504,32 @@ struct VectorizeLoops_ : public IRMutator<Expr *> {
     if (!for_min_i) return Expr();
     if (for_min_i->value != 0) return Expr();
 
-    Expr times = Div::Make(forloop->extent, make_const(factor));
+    Expr times = common::AutoSimplify(Div::Make(forloop->extent, make_const(factor)));
     Simplify(&times);
 
     // update the current forloop
-    forloop->extent = times;
+    auto times_int = times.As<IntImm>();
```
Comment (Superjomn) [inline cinn/optim/vectorize_loops.cc:483]: times.as_int()

File: cinn/optim/vectorize_loops.cc [hunk 7]
```diff
@@ -277,19 +303,97 @@ class Vectorizer : public IRMutator<Expr *> {
...
+
   void Visit(const For *forloop, Expr *expr) {
     auto *node = expr->As<For>();
+    if (forloop->extent.As<IntImm>()) {
+      var_intervals.emplace(forloop->loop_var->name, common::CasInterval{0, forloop->extent.as_int32() - 1});
+    } else {
+      var_intervals.emplace(forloop->loop_var->name, common::CasInterval{Expr(0), forloop->extent - 1});
+    }
+    // // unroll for if it has cinn_min extent(has tail block) by solve the condition and then vectorize separately
+    // Block *block = node->body.As<Block>();
```
Comment (Superjomn) [inline cinn/optim/vectorize_loops.cc]: clean useless code

File: cinn/common/ir_util.cc (hunk 1 above)
Comment (wenming2014) [reply cinn/common/ir_util.cc]: done,thanks~

File: cinn/hlir/framework/buffer.h (hunk 2 above)
Comment (wenming2014) [reply cinn/hlir/framework/buffer.h]: done,thanks~

File: cinn/hlir/pe/schedule.cc (hunk 3 above)
Comment (wenming2014) [reply cinn/hlir/pe/schedule.cc:43]: schedule都规整到schedule.cc里了

File: cinn/optim/vectorize_loops.cc (hunk 6 above)
Comment (wenming2014) [reply cinn/optim/vectorize_loops.cc:483]: extent不一定是int，有可能是min/max

File: tests/benchmark/test_all_ops_default.cc (hunk 5 above)
Comment (wenming2014) [reply tests/benchmark/test_all_ops_default.cc]: done,thanks~

File: cinn/hlir/pe/schedule.cc (hunk 3 above)
Comment (haozech) [reply cinn/hlir/pe/schedule.cc:43]: 这里是整理了代码，从nn.cc移到了schedule.cc

File: cinn/common/cas.cc [hunk 8]
```diff
@@ -955,7 +1335,79 @@ Expr CasSimplifyMutator::operator()(Expr u) {
   }
 
   if (u.As<Sum>()) {
-    return detail::SumOrProductGetSingleElementsRec(SimplifySum(u));
+    auto tmp = detail::SumOrProductGetSingleElementsRec(SimplifySum(u));
+    // deal with index's div-mod add simplification, tempory solution, not cover all situations.
+    // case 1: m / n * n + m % n = m (m, n's type is int)
+    // case 2: m / n1 * n3 + n2 * m % n3 = n2 * m if n3 = n1 * n2 (m, n1, n2, n3's type is int)
+    // case 3: m / n2 + n1 * m % (n3) = n1 * m if n3 = n1 * n2 (m, n1, n2, n3's type is int)
```
Comment (haozech) [inline cinn/common/cas.cc:1419]: 多行注释可以考虑用
/*
 \* ...
 \* ...
 */
的格式，把注释放在函数名前作为函数的详细说明。

File: cinn/common/cas_test.cc [hunk 9]
```diff
@@ -314,7 +314,112 @@ TEST(SolveInequality, basic) {
...
+    LOG(INFO) << "simplified " << p2;
+    #ifdef CINN_WITH_CUDA
+      EXPECT_EQ(GetStreamCnt(p2), "((32 + ((-1 * x1) + (-16 * x2))) % 33)");
+    #else
+      EXPECT_EQ(GetStreamCnt(p2), "(32 + (((-1 * x1) + (-16 * x2)) % 33))");
+    #endif
+  }
+}
+TEST(CAS, SimplifyNegtive) {
+  {  // bug: (-1*x) /2
```
Comment (haozech) [inline cinn/common/cas_test.cc]: 是否是需要删除的注释？

File: cinn/ir/ir_operators.cc [hunk 10]
```diff
@@ -1,12 +1,14 @@
...
 
 #include <limits>
+#include <string>
 
 #include "cinn/common/type.h"
 #include "cinn/lang/compute.h"
 
 namespace cinn {
 namespace ir {
+using attr_t = std::variant<int, float, bool, std::string>;
```
Comment (haozech) [inline cinn/ir/ir_operators.cc:11]: 这里和cinn/ir/ir.h#L35的语句相同，感觉可以改为只在一个地方定义attr_t，其余地方直接using ir::attr_t

File: cinn/lang/compute.h [hunk 11]
```diff
@@ -13,6 +14,7 @@ namespace cinn {
 namespace lang {
 
 using compute_handler_t = std::function<Expr(const std::vector<Expr> &)>;
+using attr_t            = std::variant<int, float, bool, std::string>;
```
Comment (haozech) [inline cinn/lang/compute.h:17]: 同上，避免多次定义

**Review by haozech — APPROVED** [review]
LGTM

**Pull request merged**